### PR TITLE
Don't build unnecessary jars.

### PR DESCRIPTION
**Pull request by nkoroste**
# Checklist

- [x] I have filed an issue about this change and discussed potential changes with the maintainers.
- [ ] I have received the approval from the maintainers to make this change.
- [x] This is not a stylistic, refactoring, or cleanup change.

Please note that the maintainers will not be reviewing this change until all checkboxes are ticked. See 
the [Contributions](https://github.com/bazelbuild/intellij#contributions) section in the README for more 
details.

# Discussion thread for this change

Issue number: https://github.com/bazelbuild/intellij/issues/1785

# Description of this change

reopened https://github.com/bazelbuild/intellij/pull/1786 - Don't build the class jar if source jars are available, as we can use that in conjunction with the interface jar to index the code instead.

For our app it reduces `~30%` of the Jars the IDE will index (from `14484` to `10072`)

As per PR review request I put this feature behind `optimize_building_jars` experiment which is enabled by default. In case an issue is found with this feature this experiment can be disabled by disabled it in the IntelliJ experiments file as follows:
```
echo optimize_building_jars=0 >> ~/.intellij-experiments
```

### CODE REVIEW

**Comment by agluszak** [comment]
Hi! 

> For our app it reduces ~30% of the Jars the IDE will index (from 14484 to 10072)
Would you be willing to run some benchmarks on open source projects and provide the results here?

File: aspect/intellij_info_impl.bzl [hunk 1]
```diff
@@ -188,9 +188,11 @@ def jars_from_output(output):
     """Collect jars for intellij-resolve-files from Java output."""
     if output == None:
         return []
+    source_jars = get_source_jars(output)
+    jars = source_jars
```
Comment (blorente) [inline aspect/intellij_info_impl.bzl:192]: What is this assignment for?

**Review by blorente — APPROVED** [review]
The implementation looks good to me, and I think it's unlikely to interfere with indexing. However, as @agluszak said, given that this changes a lot of the core resolution, it would be interesting to run the same benchmark in other projects (such as this repository, for instance) before merging, to make sure it's not a regression.

File: aspect/intellij_info_impl.bzl (hunk 1 above)
Comment (nkoroste) [reply aspect/intellij_info_impl.bzl:192]: Not needed, you're right. 
```suggestion
```

**Comment by nkoroste** [comment]
> Hi!
> 
> > For our app it reduces ~30% of the Jars the IDE will index (from 14484 to 10072)
> 
> Would you be willing to run some benchmarks on open source projects and provide the results here?

and 

> The implementation looks good to me, and I think it's unlikely to interfere with indexing. However, as @agluszak said, given that this changes a lot of the core resolution, it would be interesting to run the same benchmark in other projects (such as this repository, for instance) before merging, to make sure it's not a regression.

I can provide a rough count of how many Jars would be ignored however I don't know how interesting that number is, instead what would be more interesting is to measure the impact this may have on indexing and code-completion time but unfortunately Intellij  IDE itself nor this plugin provide an easy way to get those numbers

**Comment by mai93** [comment]
Can we try to incrementally roll it out?, I feel like we have been adding a lot of feature flags and not all of them are documented so not a lot of users get to try the features they protect.

**Comment by mai93** [comment]
We can use [`FeatureRolloutExperiment`](https://github.com/bazelbuild/intellij/blob/master/common/experiments/src/com/google/idea/common/experiments/FeatureRolloutExperiment.java) to define the experiment flag and set the roll out percentage in [experiment.properties](https://github.com/bazelbuild/intellij/blob/master/common/experiments/src/com/google/idea/common/experiments/experiment.properties)

**Comment by mai93** [comment]
Follow up on how we can select between the old and new behaviour in the aspect, we can use [--aspects_parameters](https://bazel.build/reference/command-line-reference#flag--aspects_parameters) to pass a boolean value to the aspect based on the experiment value (i.e. enabled or not) and select the one of the 2 approaches in the aspect based on this boolean. We need to add a boolean attribute in the aspect definition, can be called `optimize_building_jars` and pass its value from the command line as `--aspects_parameters=optimize_building_jars=true`.

File: aspect/testing/tests/src/com/google/idea/blaze/aspect/java/dependencies/DependenciesTest.java [hunk 2]
```diff
@@ -103,20 +100,17 @@ public void testJavaLibraryWithExports() throws Exception {
...
             testRelative("libfoo_exporter-src.jar"),
-            testRelative("libexport_consumer.jar"),
             testRelative("libexport_consumer-hjar.jar"),
             testRelative("libexport_consumer-src.jar"));
     assertThat(getOutputGroupFiles(testFixture, "intellij-compile-java"))
         .containsAtLeast(
             testRelative("libfoo.jar"),
             testRelative("libfoo_exporter.jar"),
-            testRelative("libexport_consumer.jar"));
+            testRelative("libexport_consumer.jar"));z
```
Comment (nkoroste) [inline aspect/testing/tests/src/com/google/idea/blaze/aspect/java/dependencies/DependenciesTest.java:113]:
```suggestion
            testRelative("libexport_consumer.jar"));
```

**Comment by nkoroste** [comment]
PTAL @mai93 @tpasternak

File: aspect/testing/tests/src/com/google/idea/blaze/aspect/java/javabinary/JavaBinaryTest.java [hunk 3]
```diff
@@ -58,10 +58,8 @@ public void testJavaBinary() throws Exception {
             testRelative(intellijInfoFileName("foo")));
     assertThat(getOutputGroupFiles(testFixture, "intellij-resolve-java"))
         .containsAtLeast(
-            testRelative("libfoolib.jar"),
```
Comment (mai93) [inline aspect/testing/tests/src/com/google/idea/blaze/aspect/java/javabinary/JavaBinaryTest.java:61]: why these jars are no longer expected, is optimization enabled for this test? same for `DependenciesTest`

Comment (nkoroste) [reply aspect/testing/tests/src/com/google/idea/blaze/aspect/java/javabinary/JavaBinaryTest.java]: Reverted these changes, they were traces to when optimization was enabled by default.

**Comment by mai93** [comment]
it looks good to me, I can merge it if none else wants to review cc @tpasternak @agluszak

**Comment by aschott-looker** [comment]
This option isn't supported on the version of Bazel we use. How do you disable it for a project?

**Comment by mai93** [comment]
what feature is not supported in the Bazel version you use?

@nkoroste mentioned that this experiment can be disabled by disabled it in the IntelliJ experiments file as follows:

`echo optimize_building_jars=0 >> ~/.intellij-experiments`

**Comment by aschott-looker** [comment]
Any Bazel version older than when [this commit](https://github.com/bazelbuild/bazel/commit/37710728712e00e8bd7145662668d72591204146) landed will fail:

```
ERROR: --aspects_parameters=optimize_building_jars=enabled :: Unrecognized option: --aspects_parameters=optimize_building_jars=enabled
```

**Comment by tpasternak** [comment]
@mai93 , @ujohnny do we still support Bazel 5?

**Comment by tpasternak** [comment]
Bazel 5 will be EOL in September. https://blog.bazel.build/2020/11/10/long-term-support-release.html

@aschott-looker are there any chances you could upgrade to bazel 6?

**Comment by hisener** [comment]
https://bazel.build/release says it's EOL in Jan 2025.

**Comment by aschott-looker** [comment]
@tpasternak We definitely want to upgrade to a more recent version of Bazel but it's not feasible in the short term. So I was hoping for a quick workaround.

Thanks for creating the issues above!

**Comment by tpasternak** [comment]
the hotfix is on its way #6653

**Pull request merged**